qgsvectorlayerchunkloader: Fix rayIntersection

ptitjano opened the pull request:
## Description

The coordinates of vector data chunks are now independant from the
scene origin. However, the `rayIntersection` does not take this change
into account. This means that the ray used to test the intersection is
expressed in the scene origin coordinates while the vector data
geometry are relative to the chunk origin. Therefore, the intersection
cannot be checked properly and always returns false.

This issue is fixed by retrieving the chunk origin based on the node
entity translation. Then, this translation is added to the transform
matrix  used to convert the geometry coordinates into the scene
coordinate system.

With this change, the different objects (the ray and the triangles)
are now expressed in the scene coordinate system and the intersection
can be properly checked. 

Fixes: 3cde17724e2703183e6c08662380d912e6665b21

ptitjano committed:
qgsvectorlayerchunkloader: Fix rayIntersection

The coordinates of vector data chunks are now independant from the
scene origin. However, the `rayIntersection` does not take this change
into account. This means that the ray used to test the intersection is
expressed in the scene origin coordinates while the vector data
geometry are relative to the chunk origin. Therefore, the intersection
cannot be checked properly and always returns false.

This issue is fixed by retrieving the chunk origin based on the node
entity translation. Then, this translation is added to the transform
matrix  used to convert the geometry coordinates into the scene
coordinate system.

With this change, the different objects (the ray and the triangles)
are now expressed in the scene coordinate system and the intersection
can be properly checked.

Fixes: 3cde17724e2703183e6c08662380d912e6665b21

wonder-sk commented on src/3d/qgsvectorlayerchunkloader_p.cpp:
```diff
@@ -317,8 +317,13 @@ QVector<QgsRayCastingUtils::RayHit> QgsVectorLayerChunkedEntity::rayIntersection
+        // This translation is stored in the QTransform component
+        // this needs to be taken into account to get the whole transformation
+        const QMatrix4x4 nodeTransformMatrix = node->entity()->findChild<Qt3DCore::QTransform *>()->matrix();
+        const QMatrix4x4 fullTransformMatrix = nodeTransformMatrix * transformMatrix;
```
The `transformMatrix` is applied after `nodeTransformMatrix`, so their order should be reversed... right now it does not matter given that both are translations, but if this changes, another bug would surface if the order is not correct:
```suggestion
        const QMatrix4x4 fullTransformMatrix = transformMatrix * nodeTransformMatrix;
```

ptitjano replied:
Done.

wonder-sk commented on src/3d/qgsvectorlayerchunkloader_p.cpp:
```diff
@@ -317,8 +317,13 @@ QVector<QgsRayCastingUtils::RayHit> QgsVectorLayerChunkedEntity::rayIntersection
+        // this needs to be taken into account to get the whole transformation
+        const QMatrix4x4 nodeTransformMatrix = node->entity()->findChild<Qt3DCore::QTransform *>()->matrix();
+        const QMatrix4x4 fullTransformMatrix = nodeTransformMatrix * transformMatrix;
         // TODO: use also geo transform matrix here???
```
geo transform matrix is now being applied, so this to-do can be removed

ptitjano replied:
Done.

wonder-sk commented on tests/src/3d/testqgs3drendering.cpp:
```diff
@@ -2577,6 +2578,71 @@ void TestQgs3DRendering::test3DSceneExporterBig()
+
+  Qgs3DUtils::captureSceneImage( engine, scene );
+  QImage img = Qgs3DUtils::captureSceneImage( engine, scene );
+  QGSVERIFYIMAGECHECK( "ray_3d", "ray_3d", img, QString(), 50, QSize( 0, 0 ), 2 );
```
why are we doing a render check here? IMHO there shouldn't be a rendering check, it just adds unnecessary test maintenance...

ptitjano replied:
I don't agree. If the test fails, one need to be able to know if this comes from a regression in the rendering or the picking.

wonder-sk replied:
But ray intersection has nothing to do with rendering... I am not sure if I understand your point.

Only now I have realized the test is added to testqgs3drendering.cpp, while in fact it should go to testqgs3dutils.cpp (or to testqgs3dmapscene.cpp) given that it is Qgs3DUtils::castRay() method that gets tested.

ptitjano replied:
Ray intersection also depends on the geometry of the layers, the camera etc.. This is what I meant by rendering. The idea is to distinguish between: the test fails but the scene is completely wrong (because of the camera, geometry, etc...) or the scene looks ok but the ray intersection still fails. This would help narrow down the origin of a regression in the code.

I wrote the test in rendering because to avoid some duplication in layers definition. Even if one could only  test `Qgs3DUtils::castRay`, one still need to define layers. This is already done in rendering but not in utils tests.

ptitjano replied:
> If the scene is completely wrong, there will be 10 other rendering tests failing, so I don't think we need to worry about the first case. I am worried that with that logic, we would be inserting loads of test images without a good reason, and then the maintenance becomes a pain.

But This issue wouldn’t be apparent if we only ran the `Qgs3DUtils` unit test. Moreover, all other scene tests use a distant camera view, while this one uses a close-up view. From my perspective, that helps improve unit test coverage.

I must admit I don’t quite understand how trying to improve test coverage in parts of the code that has regressed multiple times in the past could be perceived negatively, even if it may lead to a small increase in maintenance effort.

In any case, I don’t want to dwell on this. I’ve moved the unit test to `TestQgs3DUtils` as requested. I had to add a call to `Qgs3DUtils::captureSceneImage`, otherwise the building layer nodes weren’t loading — which might suggest that it somewhat depends on rendering.

wonder-sk replied:
> But This issue wouldn’t be apparent if we only ran the Qgs3DUtils unit test.

CI runs all tests, so I don't see this as an issue

> I must admit I don’t quite understand how trying to improve test coverage in parts of the code that has regressed multiple times in the past could be perceived negatively, even if it may lead to a small increase in maintenance effort.

From my point of view, maintaining lots of rendering tests is a pain, not a small increase in maintenance effort - they easily break with some sub-pixel shifts that are otherwise not perceivable, and images need adjusting.

> In any case, I don’t want to dwell on this. I’ve moved the unit test to TestQgs3DUtils as requested.

Thanks.

> I had to add a call to Qgs3DUtils::captureSceneImage, otherwise the building layer nodes weren’t loading

That's because the scene needs to be fully loaded first - you could as well just wait for `scene->sceneState() == Qgs3DMapScene::Ready` (that's what captureSceneImage() does internally before requesting a capture)

ptitjano commented on src/3d/qgsvectorlayerchunkloader_p.cpp:
```diff
@@ -317,8 +317,13 @@ QVector<QgsRayCastingUtils::RayHit> QgsVectorLayerChunkedEntity::rayIntersection
+        // the node geometry has been translated by chunkOrigin
+        // This translation is stored in the QTransform component
+        // this needs to be taken into account to get the whole transformation
+        const QMatrix4x4 nodeTransformMatrix = node->entity()->findChild<Qt3DCore::QTransform *>()->matrix();
```
Done.

ptitjano committed:
testqgs3dutils: Init Qgs3d

This will make it possible to use the 3D symbol types in the following
commits.

ptitjano committed:
testqgs3dutils: Store rgb layer

This layer will be reused in the next commit.
